pkg/webhook/generator.go

```diff
@@ -0,0 +1,334 @@
+		CABundle: []byte{},
+	}
+	if o.host != nil {
+		u := url.URL{
+			Scheme: "https",
+			Host:   net.JoinHostPort(*o.host, strconv.Itoa(int(o.port))),
+		}
+		urlString := u.String()
+		cc.URL = &urlString
+	}
+	if o.service != nil {
+		cc.Service = &admissionregistration.ServiceReference{
+			Name:      o.service.name,
+			Namespace: o.service.namespace,
+			// Path will be set later
+		}
+	}
+	return cc, nil
+}
+
+// getClientConfigWithPath constructs a WebhookClientConfig based on the server generatorOptions.
+// It will use path to the set the path in WebhookClientConfig.
+func (o *generatorOptions) getClientConfigWithPath(path string) (*admissionregistration.WebhookClientConfig, error) {
+	cc, err := o.getClientConfig()
+	if err != nil {
+		return nil, err
+	}
+	return cc, setPath(cc, path)
+}
+
+// setPath sets the path in the WebhookClientConfig.
+func setPath(cc *admissionregistration.WebhookClientConfig, path string) error {
+	if cc.URL != nil {
+		u, err := url.Parse(*cc.URL)
+		if err != nil {
+			return err
+		}
+		u.Path = path
+		urlString := u.String()
+		cc.URL = &urlString
+	}
+	if cc.Service != nil {
+		cc.Service.Path = &path
+	}
+	return nil
+}
+
+// whConfigs creates a mutatingWebhookConfiguration and(or) a validatingWebhookConfiguration based on registry.
+// For the same type of webhook configuration, it generates a webhook entry per endpoint.
+func (o *generatorOptions) whConfigs() ([]runtime.Object, error) {
+	for _, webhook := range o.registry {
+		if err := webhook.Validate(); err != nil {
+			return nil, err
+		}
+	}
+
+	objs := []runtime.Object{}
+	mutatingWH, err := o.mutatingWHConfigs()
+	if err != nil {
+		return nil, err
+	}
+	if mutatingWH != nil {
+		objs = append(objs, mutatingWH)
+	}
+	validatingWH, err := o.validatingWHConfigs()
+	if err != nil {
+		return nil, err
+	}
+	if validatingWH != nil {
+		objs = append(objs, validatingWH)
+	}
+	return objs, nil
+}
+
+func (o *generatorOptions) mutatingWHConfigs() (runtime.Object, error) {
+	mutatingWebhooks := []v1beta1.Webhook{}
+	for path, webhook := range o.registry {
+		if webhook.GetType() != webhookTypeMutating {
+			continue
+		}
+
+		admissionWebhook := webhook.(*admissionWebhook)
+		wh, err := o.admissionWebhook(path, admissionWebhook)
+		if err != nil {
+			return nil, err
+		}
+		mutatingWebhooks = append(mutatingWebhooks, *wh)
+	}
+
+	sort.Slice(mutatingWebhooks, func(i, j int) bool {
+		return mutatingWebhooks[i].Name < mutatingWebhooks[j].Name
+	})
+
+	if len(mutatingWebhooks) > 0 {
+		return &admissionregistration.MutatingWebhookConfiguration{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: metav1.GroupVersion{Group: admissionregistration.GroupName, Version: "v1beta1"}.String(),
+				Kind:       "MutatingWebhookConfiguration",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: o.mutatingWebhookConfigName,
+				Annotations: map[string]string{
+					"admissionwebhook.alpha.kubebuilder.io/ca-secret-name": "webhook-cert",
```

Comment:
Two comments:
 - We should add a comment explaining (or linking to an issue/doc) why this annotation so readers understand it.
 - annotation name "alpha" needs to come before admissionwebhooks ? `alpha.admissionwebhooks.kubebuilder.io/ca-secret-name`
 - "webhook-cert" is hardcoded ?

Reply:
I now uses `alpha.admissionwebhook.kubebuilder.io/ca-secret-name`.

> "webhook-cert" is hardcoded ?

fixed.